usr/src/man/man8/pptadm.8
.\"
.\" This file and its contents are supplied under the terms of the
.\" Common Development and Distribution License ("CDDL"), version 1.0.
.\" You may only use this file in accordance with the terms of version
.\" 1.0 of the CDDL.
.\"
.\" A full copy of the text of the CDDL should have accompanied this
.\" source.  A copy of the CDDL is also available via the Internet at
.\" http://www.illumos.org/license/CDDL.
.\"
.\" Copyright 2018 Joyent, Inc.
.\" Copyright 2024 MNX Cloud, Inc.
.\"
.Dd October 16, 2024
.Dt PPTADM 8
.Os
.Sh NAME
.Nm pptadm
.Nd PPT administration utility
.Sh SYNOPSIS
.Nm
.Cm list -j Op Fl a Op Fl UN
.Nm
.Cm list
.Op Fl ap Op Fl UN
.Op Fl o Ar fields
.Sh DESCRIPTION
The
.Nm
utility can enumerate passthrough devices for use by a virtualized guest.
.Sh OPTIONS
The following options to the
.Cm list
command are supported:
.Bl -tag -width Ds
.It Fl a
Show all PPT devices, both available and assigned.
.It Fl j
Output JSON.
.It Fl o
Specify fields to output, or "all". Available fields are
dev,path,vendor,device,subvendor,subdevice,rev,label
.It Fl p
Output in a parsable format; this requires the -o option to be specified.
.It Fl N
Output native device driver names on available devices if applicable; this
requires the -a option to be specified.
.It Fl U
Output unavailable devices that have no native device driver, for use in
populating /etc/ppt_matches (see below). This requires the -a option to be
specified.
.El
.Sh JSON OUTPUT
The JSON output consists of an array under the key "devices" with the fields:
.Bl -tag -width Ds
.It dev
The PPT /dev path, if assigned and bound.
.It path
The physical /devices path.
.It vendor-id
The PCI vendor ID.
.It device-id
The PCI device ID.
.It subsystem-vendor-id
The PCI subsystem vendor ID.
.It subsystem-id
The PCI subsystem ID.
.It revision-id
The PCI device revision.
.It label
Human-readable description from the PCI database.
.El
.Sh FILES
.Bl -tag -width Ds
.It /etc/ppt_aliases
Containts the bindings of PPT devices in the same format as /etc/driver_aliases
.It /etc/ppt_matches
Identifies devices that PPT could be bound to, either by physical path, or by
the device's PCI ID.
.El
.Sh EXIT STATUS
.Ex -std


usr/src/cmd/pptadm/pptadm.c
/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 *
 * Copyright 2018 Joyent, Inc.
 * Copryight 2024 MNX Cloud, Inc.
 */

#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>
#include <getopt.h>
#include <string.h>
#include <ofmt.h>
#include <err.h>
#include <zone.h>

#include <libppt.h>

typedef enum field {
	PPT_DEV,
	PPT_VENDOR,
	PPT_DEVICE,
	PPT_SUBVENDOR,
	PPT_SUBDEVICE,
	PPT_REV,
	PPT_PATH,
	PPT_LABEL
} field_t;

const char *valname[] = {
	"dev",
	"vendor-id",
	"device-id",
	"subsystem-vendor-id",
	"subsystem-id",
	"revision-id",
	"path",
	"label"
};

static ofmt_cb_t print_field;

static ofmt_field_t fields[] = {
/* name,	field width, index, callback */
{ "DEV",	sizeof ("/dev/pptXX"), PPT_DEV, print_field },
{ "VENDOR",	sizeof ("VENDOR"), PPT_VENDOR, print_field },
{ "DEVICE",	sizeof ("DEVICE"), PPT_DEVICE, print_field },
{ "SUBVENDOR",	sizeof ("SUBVENDOR"), PPT_SUBVENDOR, print_field },
{ "SUBDEVICE",	sizeof ("SUBDEVICE"), PPT_SUBDEVICE, print_field },
{ "REV",	sizeof ("REV"), PPT_REV, print_field },
{ "PATH",	50, PPT_PATH, print_field },
{ "LABEL",	60, PPT_LABEL, print_field },
{ NULL,		0, 0, NULL },
};

/* For some outputs we want wider device paths. */
#define	devwidth (fields[0].of_width)

static void
usage(const char *errmsg)
{
	if (errmsg != NULL)
		(void) fprintf(stderr, "pptadm: %s\n", errmsg);
	(void) fprintf(errmsg != NULL ? stderr : stdout,
	    "Usage:\n"
	    "pptadm list [ -j ] [ -a [-UN]]\n"
	    "pptadm list [-ap [-UN]] [-o fields]\n");
	exit(errmsg != NULL ? EXIT_FAILURE : EXIT_SUCCESS);
}

/* PRINTFLIKE1 */
static void
die(const char *fmt, ...)
{
	va_list ap;
	va_start(ap, fmt);
	verrx(EXIT_FAILURE, fmt, ap);
	va_end(ap);
}

/* Has to use boolean_t because ofmt_cb_t needs it. */
static boolean_t
print_field(ofmt_arg_t *arg, char *buf, uint_t bufsize)
{
	nvlist_t *nvl = arg->ofmt_cbarg;
	nvpair_t *nvp = NULL;

	while ((nvp = nvlist_next_nvpair(nvl, nvp)) != NULL) {
		const char *name = nvpair_name(nvp);
		char *val = NULL;

		(void) nvpair_value_string(nvp, &val);

		if (strcmp(name, valname[arg->ofmt_id]) != 0)
			continue;

		(void) snprintf(buf, bufsize, "%s", val);
		return (B_TRUE);
	}

	(void) snprintf(buf, bufsize, "--");
	return (B_TRUE);
}

static int
list(int argc, char *argv[])
{
	const char *fields_str = NULL;
	bool parsable = false;
	bool json = false;
	bool all = false;
	bool unknown = false;
	bool nativepath = false;
	uint_t ofmtflags = 0;
	ofmt_status_t oferr;
	ofmt_handle_t ofmt;
	int opt;

	while ((opt = getopt(argc, argv, "ahjo:pNU")) != -1) {
		switch (opt) {
		case 'a':
			all = true;
			break;
		case 'h':
			usage(NULL);
			break;
		case 'j':
			json = true;
			break;
		case 'o':
			fields_str = optarg;
			break;
		case 'p':
			ofmtflags |= OFMT_PARSABLE;
			parsable = true;
			break;
		case 'N':
			devwidth = 20;
			nativepath = true;
			break;
		case 'U':
			unknown = true;
			break;
		default:
			usage("unrecognized option");
			break;
		}
	}

	if (!all && (nativepath || unknown))
		usage("-N and/or -U cannot be used without -a");

	if (optind == (argc - 1))
		usage("unused arguments");

	if (json && (parsable || fields_str != NULL))
		usage("-j option cannot be used with -p or -o options");

	if (fields_str == NULL) {
		if (parsable)
			usage("-o must be provided when using -p option");
		fields_str = "dev,vendor,device,path";
	}

	oferr = ofmt_open(fields_str, fields, ofmtflags, 0, &ofmt);

	ofmt_check(oferr, parsable, ofmt, die, warn);

	nvlist_t *nvl = all ? ppt_list(unknown, nativepath) :
	    ppt_list_assigned();
	nvpair_t *nvp = NULL;

	if (json) {
		if (printf("{\n\t\"devices\": [\n") < 0)
			err(EXIT_FAILURE, "failed to write JSON");
	}

	while ((nvp = nvlist_next_nvpair(nvl, nvp)) != NULL) {
		nvlist_t *props;

		(void) nvpair_value_nvlist(nvp, &props);

		if (json) {
			if (printf("\t\t") < 0)
				err(EXIT_FAILURE, "failed to write JSON");
			if (nvlist_print_json(stdout, props) < 0)
				err(EXIT_FAILURE, "failed to write JSON");
			if (nvlist_next_nvpair(nvl, nvp) != NULL)
				(void) printf(",\n");
		} else {
			ofmt_print(ofmt, props);
		}
	}

	if (json) {
		if (printf("\n\t]\n}\n") < 0)
			err(EXIT_FAILURE, "failed to write JSON");
	}

	nvlist_free(nvl);
	ofmt_close(ofmt);
	return (EXIT_SUCCESS);
}

int
main(int argc, char *argv[])
{
	/* Don't even bother unless we're in the global zone. */
	if (getzoneid() != GLOBAL_ZONEID)
		usage("pptadm(8) can only be used in the global zone.");

	if (argc == 1)
		return (list(argc - 1, argv));

	if (strcmp(argv[1], "list") == 0) {
		return (list(argc - 1, &argv[1]));
	} else {
		usage("unknown sub-command");
	}

	return (EXIT_SUCCESS);
}
